### ISSUE TITLE
Daemon fails due to too many open file handlers

### ISSUE THREAD

**Issue by billbitt** [lbryio/lbry#810]
## The Issue
I am running the daemon on a linux server (spee.ch).  From time to time, the daemon becomes unresponsive to certain requests, such as 'get' and 'resolve' requests.

### Steps to reproduce
I do not know how to reproduce exactly; it happens intermittently, likely due to large traffic or from being open for a long time.  Once it starts happening, the steps to reproduce these specific errors are:
1. Run a curl command to ask the daemon to 'get' or 'resolve' a uri

### Expected behaviour
daemon should respond with the results of the 'get' or 'resolve' request

### Actual behaviour
The daemon waits and then responds with the error message: ''Failed to get response from server within timeout of 30"

## System Configuration
"id": null,
"jsonrpc": "2.0",
"result": {
  "build": "release",
  "ip": "****",
  "lbrynet_version": "0.14.1",
  "lbryschema_version": "0.0.7",
  "lbryum_version": "3.1.3",
  "os_release": "4.4.0-64-generic",
  "os_system": "Linux",
  "platform": "Linux-4.4.0-64-generic-x86_64-with-Ubuntu-16.04-xenial",
  "processor": "x86_64",
  "python_version": "2.7.12"

## Anything Else
These are the results of running the curl commands: 

```
lbry@ip-172-55-1-123:~$ curl 'http://localhost:5279/lbryapi' --data '{"method":
"wallet_balance"}'
{
  "id": null,
  "jsonrpc": "2.0",
  "result": 9.241749
}


lbry@ip-172-55-1-123:~$ curl 'http://localhost:5279/lbryapi' --data '{"method":
"get","params":{"uri":"what"} }'
{
  "error": {
    "code": -32500,
    "data": {
      "traceback": [
        "Traceback (most recent call last):",
        "  File \"site-packages/twisted/internet/defer.py\", line 498, in errbac
k",
        "    ",
        "  File \"site-packages/twisted/internet/defer.py\", line 565, in _start
RunCallbacks",
        "    ",
        "  File \"site-packages/twisted/internet/defer.py\", line 651, in _runCa
llbacks",
        "    ",
        "  File \"site-packages/twisted/internet/defer.py\", line 1355, in gotRe
sult",
        "    ",
        "--- <exception caught here> ---",
        "  File \"site-packages/twisted/internet/defer.py\", line 1297, in _inli
neCallbacks",
        "    ",
        "  File \"site-packages/twisted/python/failure.py\", line 389, in throwE
xceptionIntoGenerator",
        "    ",
        "  File \"site-packages/lbrynet/daemon/Daemon.py\", line 1481, in jsonrp
c_get",
        "    ",
        "  File \"site-packages/twisted/internet/defer.py\", line 1297, in _inli
neCallbacks",
        "    ",
        "  File \"site-packages/twisted/python/failure.py\", line 389, in throwE
xceptionIntoGenerator",
        "    ",
        "  File \"site-packages/lbrynet/core/Wallet.py\", line 815, in resolve",

        "    ",
        "  File \"site-packages/twisted/python/threadpool.py\", line 246, in inC
ontext",
        "    ",
        "  File \"site-packages/twisted/python/threadpool.py\", line 262, in <la
mbda>",
        "    ",
        "  File \"site-packages/twisted/python/context.py\", line 118, in callWi
thContext",
        "    ",
        "  File \"site-packages/twisted/python/context.py\", line 81, in callWit
hContext",
        "    ",
        "  File \"site-packages/lbryum/commands.py\", line 140, in func_wrapper"
,
        "    ",
        "  File \"site-packages/lbryum/commands.py\", line 1115, in getvaluesfor
uris",
        "    ",
        "  File \"site-packages/lbryum/network.py\", line 781, in synchronous_ge
t",
        "    ",
        "exceptions.BaseException: Failed to get response from server within tim
eout of 30",
        ""
      ]
    },
    "message": "Failed to get response from server within timeout of 30"
  },
  "id": null,
  "jsonrpc": "2.0"
}


lbry@ip-172-55-1-123:~$ curl 'http://localhost:5279/lbryapi' --data '{"method":
"resolve","params":{"uri":"one"} }'
{
  "error": {
    "code": -32500,
    "data": {
      "traceback": [
        "Traceback (most recent call last):",
        "  File \"site-packages/twisted/internet/defer.py\", line 498, in errbac
k",
        "    ",
        "  File \"site-packages/twisted/internet/defer.py\", line 565, in _start
RunCallbacks",
        "    ",
        "  File \"site-packages/twisted/internet/defer.py\", line 651, in _runCa
llbacks",
        "    ",
        "  File \"site-packages/twisted/internet/defer.py\", line 1355, in gotRe
sult",
        "    ",
        "--- <exception caught here> ---",
        "  File \"site-packages/twisted/internet/defer.py\", line 1297, in _inli
neCallbacks",
        "    ",
        "  File \"site-packages/twisted/python/failure.py\", line 389, in throwE
xceptionIntoGenerator",
        "    ",
        "  File \"site-packages/lbrynet/daemon/Daemon.py\", line 1432, in jsonrp
c_resolve",
        "    ",
        "  File \"site-packages/twisted/internet/defer.py\", line 1297, in _inli
neCallbacks",
        "    ",
        "  File \"site-packages/twisted/python/failure.py\", line 389, in throwE
xceptionIntoGenerator",
        "    ",
        "  File \"site-packages/lbrynet/core/Wallet.py\", line 815, in resolve",

        "    ",
        "  File \"site-packages/twisted/python/threadpool.py\", line 246, in inC
ontext",
        "    ",
        "  File \"site-packages/twisted/python/threadpool.py\", line 262, in <la
mbda>",
        "    ",
        "  File \"site-packages/twisted/python/context.py\", line 118, in callWi
thContext",
        "    ",
        "  File \"site-packages/twisted/python/context.py\", line 81, in callWit
hContext",
        "    ",
        "  File \"site-packages/lbryum/commands.py\", line 140, in func_wrapper"
,
        "    ",
        "  File \"site-packages/lbryum/commands.py\", line 1115, in getvaluesfor
uris",
        "    ",
        "  File \"site-packages/lbryum/network.py\", line 781, in synchronous_ge
t",
        "    ",
        "exceptions.BaseException: Failed to get response from server within tim
eout of 30",
        ""
      ]
    },
    "message": "Failed to get response from server within timeout of 30"
  },
  "id": null,
  "jsonrpc": "2.0"
}

lbry@ip-172-55-1-123:~$ curl 'http://localhost:5279/lbryapi' --data '{"method":
"status"}'
{
  "id": null,
  "jsonrpc": "2.0",
  "result": {
    "blockchain_status": {
      "best_blockhash": "f89de888dc573998f7b34e96cecb4e060388458e33ad1ab9fcac6de
477a65897",
      "blocks": 214345,
      "blocks_behind": 0
    },
    "blocks_behind": 0,
    "connection_status": {
      "code": "connected",
      "message": "No connection problems detected"
    },
    "installation_id": "*****",
    "is_first_run": false,
    "is_running": true,
    "lbry_id": "*****",
    "startup_status": {
      "code": "started",
      "message": "Started lbrynet"
    }
  }
}
```
Here are the logs from lbrynet.log:
```
2017-07-28 18:17:09,898 CRITICAL twisted:154:
ConnectionError: HTTPSConnectionPool(host='logs-01.loggly.com', port=443): Max r
etries exceeded with url: /inputs/adf4e86a-6904-4b36-97f2-0fa837d8dc0b/tag/lbryn
et-0.14.1 (Caused by NewConnectionError('<requests.packages.urllib3.connection.V
erifiedHTTPSConnection object at 0x7f9530969190>: Failed to establish a new conn
ection: [Errno 24] Too many open files',))
2017-07-28 18:17:09,899 CRITICAL twisted:154: Unhandled error in Deferred:
2017-07-28 18:17:09,899 CRITICAL twisted:154:
ConnectionError: HTTPSConnectionPool(host='logs-01.loggly.com', port=443): Max r
etries exceeded with url: /inputs/adf4e86a-6904-4b36-97f2-0fa837d8dc0b/tag/lbryn
et-0.14.1 (Caused by NewConnectionError('<requests.packages.urllib3.connection.V
erifiedHTTPSConnection object at 0x7f953095b9d0>: Failed to establish a new conn
ection: [Errno 24] Too many open files',))
2017-07-28 18:17:09,900 CRITICAL twisted:154: Unhandled error in Deferred:
2017-07-28 18:17:09,900 CRITICAL twisted:154:
ConnectionError: HTTPSConnectionPool(host='logs-01.loggly.com', port=443): Max r
etries exceeded with url: /inputs/adf4e86a-6904-4b36-97f2-0fa837d8dc0b/tag/lbryn
et-0.14.1 (Caused by NewConnectionError('<requests.packages.urllib3.connection.V
erifiedHTTPSConnection object at 0x7f953095b850>: Failed to establish a new conn
ection: [Errno 24] Too many open files',))
2017-07-28 18:17:09,901 CRITICAL twisted:154: Unhandled error in Deferred:
2017-07-28 18:17:09,901 CRITICAL twisted:154:
ConnectionError: HTTPSConnectionPool(host='logs-01.loggly.com', port=443): Max r
etries exceeded with url: /inputs/adf4e86a-6904-4b36-97f2-0fa837d8dc0b/tag/lbryn
et-0.14.1 (Caused by NewConnectionError('<requests.packages.urllib3.connection.V
erifiedHTTPSConnection object at 0x7f953095b250>: Failed to establish a new conn
ection: [Errno 24] Too many open files',))
2017-07-28 18:17:09,902 CRITICAL twisted:154: Unhandled error in Deferred:
2017-07-28 18:17:09,902 CRITICAL twisted:154:
ConnectionError: HTTPSConnectionPool(host='logs-01.loggly.com', port=443): Max r
etries exceeded with url: /inputs/adf4e86a-6904-4b36-97f2-0fa837d8dc0b/tag/lbryn
et-0.14.1 (Caused by NewConnectionError('<requests.packages.urllib3.connection.V
erifiedHTTPSConnection object at 0x7f953095b650>: Failed to establish a new conn
ection: [Errno 24] Too many open files',))
2017-07-28 18:17:09,902 CRITICAL twisted:154: Unhandled error in Deferred:
2017-07-28 18:17:09,903 CRITICAL twisted:154:
ConnectionError: HTTPSConnectionPool(host='logs-01.loggly.com', port=443): Max r
etries exceeded with url: /inputs/adf4e86a-6904-4b36-97f2-0fa837d8dc0b/tag/lbryn
et-0.14.1 (Caused by NewConnectionError('<requests.packages.urllib3.connection.V
erifiedHTTPSConnection object at 0x7f953095b510>: Failed to establish a new conn
ection: [Errno 24] Too many open files',))
2017-07-28 18:17:09,903 CRITICAL twisted:154: Unhandled error in Deferred:
2017-07-28 18:17:09,904 CRITICAL twisted:154:
ConnectionError: HTTPSConnectionPool(host='logs-01.loggly.com', port=443): Max r
etries exceeded with url: /inputs/adf4e86a-6904-4b36-97f2-0fa837d8dc0b/tag/lbryn
et-0.14.1 (Caused by NewConnectionError('<requests.packages.urllib3.connection.V
erifiedHTTPSConnection object at 0x7f953095b350>: Failed to establish a new conn
ection: [Errno 24] Too many open files',))
2017-07-28 18:17:09,904 CRITICAL twisted:154: Unhandled error in Deferred:
2017-07-28 18:17:09,905 CRITICAL twisted:154:
ConnectionError: HTTPSConnectionPool(host='logs-01.loggly.com', port=443): Max r
etries exceeded with url: /inputs/adf4e86a-6904-4b36-97f2-0fa837d8dc0b/tag/lbryn
et-0.14.1 (Caused by NewConnectionError('<requests.packages.urllib3.connection.V
erifiedHTTPSConnection object at 0x7f953095b0d0>: Failed to establish a new conn
ection: [Errno 24] Too many open files',))
```

**Comment by kaykurokawa** [comment]
lsof -p PID (where PID is the process id of the daemon) should be run on a machine where this problem pops up so we can see what file handlers it has open.

**Comment by billbitt** [comment]
these are the results of running `lsof -p` for the lbrynet-daemon processes when the issue occurred yesterday.

[pid9161.docx.txt](https://github.com/lbryio/lbry/files/1190926/pid9161.docx.txt)
[pid9159.docx.txt](https://github.com/lbryio/lbry/files/1190925/pid9159.docx.txt)

**Comment by billbitt** [comment]
happened again just now.  
`lsof -p` results:
[28344.docx.txt](https://github.com/lbryio/lbry/files/1191458/28344.docx.txt)
[28346.docx.txt](https://github.com/lbryio/lbry/files/1191459/28346.docx.txt)

lbrynet logs:
[lbrynet-8-1-17.docx.txt](https://github.com/lbryio/lbry/files/1191457/lbrynet-8-1-17.docx.txt)

**Comment by kaykurokawa** [comment]
Seems like the majority of it is for port 5279 , which is the API port. 

If you are not queueing API requests, than you probably need to increase the limit per process to a number that is more in line with the concurrent users? How many concurrent users do you have?

**Comment by kauffj** [comment]
I'd guess peak spee.ch traffic has been several hundred per minute.

Yesterday when it was going down it was probably less than 100.

**Comment by kaykurokawa** [comment]
depending on if there are multiple API calls being made per user, and considering API calls can take long to finish, it seems plausible to me that @billbitt 's problem can be solved by raising the limit , 

however this doesn't look like the same problem that we saw on reflector as reflector doesn't use API calls.

**Comment by billbitt** [comment]
(recap from meeting for Grin): it would make sense if the issue has to do with API calls (I am making a lot of them, and currently rewriting spee.ch to minimize that a bit while reorganizing the url structure), however, even after peak usage drops off the daemon still gives the 'too many open' error and will not stop until I at least kill all processes and restart daemon, so that makes me think that even if the high usage is causing the problem something else is going on that the files aren't closed after usage drops off.

**Comment by lyoshenka** [comment]
Thanks @billbitt. im looking at this now. there are two major sources of open files: the API and reflector. right now reflector actually has the majority of the connections (468 of 726) while the API has 175. So I don't think its only the API's fault.

**Comment by lyoshenka** [comment]
Short-term solution: raise the limit way up. I did this on reflector and speech.

Long-term limit: daemon needs to manage its connections to reflector better. Probably a limit on how much content it can download at once. New issue filed here: https://github.com/lbryio/lbry/issues/817

**Comment by kauffj** [comment]
@billbitt experienced this yesterday. I think there is a deeper bug here.

**Comment by billbitt** [comment]
Confirmed, yes it happened again yesterday. As discussed on call today, it seems to be a result of cumulative usage rather than total usage at a given moment.

I was under the impression this particular issue was a result of heavy use of the lbrynet api (spee.ch makes a lot of resolve, get, and claim_list calls in its current form), however it sounds like the team is thinking it is related more to reflection.  As to reflection, I have recently noticed that none (it is hard to definitively say 'none', but based on my testing it is accurate) of the spee.ch uploads are being reflected.  Two examples: I am unable to download `lbry://its-working` or `lbry://hellolbry` through the app.
